### PR TITLE
docs: add best practice docs to avoid race condition bw kubelet and apisix

### LINKED ISSUE #2030: bug: request timed out when upstream pod is scaling down
### Current Behavior

I am facing request timed out when upstream pods scales down
I've tested this by running k6s against the golang hello world service exposed via APISIX, when I trigger a scale down from 3 replicas to 1 replicas, request timed out error started showing up roughly a minute later.
I've also cross tested this with ingress-nginx and the similar behaviour does not happen.

### Expected Behavior

Request timed out should not occur

### Error Logs

This might be related but I am not sure
k6 Logs:
```
WARN[0227] Request Failed                                error="Get \"https://REDACTED/staging-apigw-loadtest/\": request timeout"
WARN[0228] Request Failed                                error="Get \"https://REDACTED/staging-apigw-loadtest/\": request timeout"
WARN[0228] Request Failed                                error="Get \"https://REDACTED/staging-apigw-loadtest/\": request timeout"
WARN[0228] Request Failed                                error="Get \"https://REDACTED/staging-apigw-loadtest/\": request timeout"
WARN[0228] Request Failed                                error="Get \"https://REDACTED/staging-apigw-loadtest/\": request timeout"
WARN[0228] Request Failed                                error="Get \"https://REDACTED/staging-apigw-loadtest/\": request timeout"
WARN[0228] Request Failed                                error="Get \"https://REDACTED/staging-apigw-loadtest/\": request timeout"
WARN[0228] Request Failed                                error="Get \"https://REDACTED/staging-apigw-loadtest/\": request timeout"
WARN[0228] Request Failed                                error="Get \"https://REDACTED/staging-apigw-loadtest/\": request timeout"
WARN[0228] Request Failed                                error="Get \"https://REDACTED/staging-apigw-loadtest/\": request timeout"
WARN[0228] Request Failed                                error="Get \"https://REDACTED/staging-apigw-loadtest/\": request timeout"
WARN[0228] Request Failed                                error="Get \"https://REDACTED/staging-apigw-loadtest/\": request timeout"
WARN[0228] Request Failed                                error="Get \"https://REDACTED/staging-apigw-loadtest/\": request timeout"
WARN[0228] Request Failed                                error="Get \"https://REDACTED/staging-apigw-loadtest/\": request timeout"
WARN[0228] Request Failed                                error="Get \"https://REDACTED/staging-apigw-loadtest/\": request timeout"
WARN[0228] Request Failed                                error="Get \"https://REDACTED/staging-apigw-loadtest/\": request timeout"
WARN[0228] Request Failed                                error="Get \"https://REDACTED/staging-apigw-loadtest/\": request timeout"
WARN[0228] Request Failed                                error="Get \"https://REDACTED/staging-apigw-loadtest/\": request timeout"
WARN[0228] Request Failed                                error="Get \"https://REDACTED/staging-apigw-loadtest/\": request timeout"
WARN[0228] Request Failed                                error="Get \"https://REDACTED/staging-apigw-loadtest/\": request timeout"
WARN[0228] Request Failed                                error="Get \"https://REDACTED/staging-apigw-loadtest/\": request timeout"
WARN[0228] Request Failed                                error="Get \"https://REDACTED/staging-apigw-loadtest/\": request timeout"
WARN[0228] Request Failed                                error="Get \"https://REDACTED/staging-apigw-loadtest/\": request timeout"
WARN[0228] Request Failed                                error="Get \"https://REDACTED/staging-apigw-loadtest/\": request timeout"
WARN[0228] Request Failed                                error="Get \"https://REDACTED/staging-apigw-loadtest/\": request timeout"
WARN[0228] Request Failed                                error="Get \"https://REDACTED/staging-apigw-loadtest/\": request timeout"
WARN[0228] Request Failed                                error="Get \"https://REDACTED/staging-apigw-loadtest/\": request timeout"
WARN[0228] Request Failed                                error="Get \"https://REDACTED/staging-apigw-loadtest/\": request timeout"
WARN[0228] Request Failed                                error="Get \"https://REDACTED/staging-apigw-loadtest/\": request timeout"
WARN[0228] Request Failed                                error="Get \"https://REDACTED/staging-apigw-loadtest/\": request timeout"
WARN[0228] Request Failed                                error="Get \"https://REDACTED/staging-apigw-loadtest/\": request timeout"
WARN[0228] Request Failed                                error="Get \"https://REDACTED/staging-apigw-loadtest/\": request timeout"
WARN[0228] Request Failed                                error="Get \"https://REDACTED/staging-apigw-loadtest/\": request timeout"
WARN[0228] Request Failed                                error="Get \"https://REDACTED/staging-apigw-loadtest/\": request timeout"
WARN[0228] Request Failed                                error="Get \"https://REDACTED/staging-apigw-loadtest/\": request timeout"
WARN[0228] Request Failed                                error="Get \"https://REDACTED/staging-apigw-loadtest/\": request timeout"
WARN[0228] Request Failed                                error="Get \"https://REDACTED/staging-apigw-loadtest/\": request timeout"
```
APISIX Logs:
```
2023/11/02 08:57:40 [error] 49#49: *308877 upstream timed out (110: Connection timed out) while connecting to upstream, client: 10.2.7.175, server: _, request: "GET /staging-apigw-loadtest/ HTTP/2.0", upstream: "http://10.2.47.135:8080/", host:  REDACTED
2023/11/02 08:57:41 [error] 51#51: *309863 upstream timed out (110: Connection timed out) while connecting to upstream, client: 10.2.15.107, server: _, request: "GET /staging-apigw-loadtest/ HTTP/2.0", upstream: "http://10.2.47.135:8080/", host: REDACTED
2023/11/02 08:57:41 [error] 51#51: *309123 upstream timed out (110: Connection timed out) while connecting to upstream, client: 10.2.18.115, server: _, request: "GET /staging-apigw-loadtest/ HTTP/2.0", upstream: "http://10.2.47.135:8080/", host: REDACTED
2023/11/02 08:57:41 [error] 52#52: *308838 upstream timed out (110: Connection timed out) while connecting to upstream, client: 10.2.7.175, server: _, request: "GET /staging-apigw-loadtest/ HTTP/2.0", upstream: "http://10.2.47.135:8080/", host:  REDACTED
2023/11/02 08:57:41 [error] 51#51: *308757 upstream timed out (110: Connection timed out) while connecting to upstream, client: 10.2.7.175, server: _, request: "GET /staging-apigw-loadtest/ HTTP/2.0", upstream: "http://10.2.47.135:8080/", host:  REDACTED
```

### Steps to Reproduce

1. Create a Deployment which host a simple endpoint that returns hello world
2. Expose the deployment
3. Create an ApisixRoute which point to the service
4. Scale the Deployment up to 3 replicas
5. Run k6 against the endpoint
6. Scale the Deployment down to 1 replica
7. Wait for 1 minutes
8. Request timed out warning shows up in k6

### Environment

- APISIX Ingress controller version: v1.7.1 (compiled myself using from the v1.7.1 tag)
- Kubernetes cluster version: version.Info{Major:"1", Minor:"23+", GitVersion:"v1.23.17-eks-be83b3c", GitCommit:"d6adb24671245f68ce3cd985f6b68f124953968d", GitTreeState:"clean", BuildDate:"2023-09-27T17:22:23Z", GoVersion:"go1.19.10", Compiler:"gc", Platform:"linux/amd64"}

### PR DESCRIPTION
**Pull request by Revolyssup**
### Type of change:
Fixes #2030 

- [ ] Bugfix
- [ ] New feature provided
- [ ] Improve performance
- [ ] Backport patches
- [x] Documentation
- [ ] Refactor
- [ ] Chore
- [ ] CI/CD or Tests

### Pre-submission checklist:

- [ ] Did you explain what problem does this PR solve? Or what new features have been added?
- [ ] Have you added corresponding test cases?
- [ ] Have you modified the corresponding document?
- [ ] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix-ingress-controller#community) first**

### CODE REVIEW

File: docs/en/latest/tutorials/how-to-avoid-race-condition-when-scaling-down.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+## The race condition
+
+APISIX ingress controller also watches for pod updates(for upstream pods). And updates the APISIX instance with new upstreams and removes the older upstreams. But this process might take time depending on the latency between Ingress controller and APISIX instance. This latency can be large if APISIX and Ingress controller have a number of network hops between them or it can be low if you're running APISIX ingress controller in [composite mode](../composite.md) where both ingress controller and APISIX instance are running in the same pod. But there is some latency(t) and if the time taken by the Kubelet to terminate those pods is less than t then for some period of time users might get 5xx errors because few of the requests may get routed to the upstreams(terminated pods) that are no longer available.
+
+## Pre-stop hook
+
+If one of the Pod's containers has defined a preStop hook and the terminationGracePeriodSeconds in the Pod spec is not set to 0, the kubelet runs that hook inside of the container. The default terminationGracePeriodSeconds setting is 30 seconds.
```
Comment (kayx23) [inline docs/en/latest/tutorials/how-to-avoid-race-condition-when-scaling-down.md:35]:
```suggestion
If one of the Pod's containers has defined a `PreStop` hook and the `terminationGracePeriodSeconds` in the Pod spec is not set to 0, the kubelet runs that hook inside of the container. The default `terminationGracePeriodSeconds` setting is 30 seconds.
```

File: docs/en/latest/tutorials/how-to-avoid-race-condition-when-scaling-down.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+## Solution
+
+You can add delay to the termination of the upstream pod by some seconds (for eg: 5 seconds) to make sure that the pod is not terminated before the APISIX instance gets updated.
+
+:::note
+
+If the preStop hook needs longer to complete than the default grace period allows, you must modify terminationGracePeriodSeconds to suit this. This is 30 seconds by default. For more information, refer to the [Kubernetes docs](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-termination).
```
Comment (kayx23) [inline docs/en/latest/tutorials/how-to-avoid-race-condition-when-scaling-down.md:43]:
```suggestion
If the preStop hook needs longer to complete than the default grace period allows, you must modify `terminationGracePeriodSeconds` to suit this. This is 30 seconds by default. For more information, refer to the [Kubernetes docs](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-termination).
```

File: docs/en/latest/tutorials/how-to-avoid-race-condition-when-scaling-down.md [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+Pods that shut down slowly should not continue to serve regular traffic and should start terminating and finish processing open connections.
+
+## The race condition
+
+APISIX ingress controller also watches for pod updates(for upstream pods). And updates the APISIX instance with new upstreams and removes the older upstreams. But this process might take time depending on the latency between Ingress controller and APISIX instance. This latency can be large if APISIX and Ingress controller have a number of network hops between them or it can be low if you're running APISIX ingress controller in [composite mode](../composite.md) where both ingress controller and APISIX instance are running in the same pod. But there is some latency(t) and if the time taken by the Kubelet to terminate those pods is less than t then for some period of time users might get 5xx errors because few of the requests may get routed to the upstreams(terminated pods) that are no longer available.
+
+## Pre-stop hook
```
Comment (kayx23) [inline docs/en/latest/tutorials/how-to-avoid-race-condition-when-scaling-down.md:33]:
```suggestion
## PreStop hook
```

**Review by kayx23 — COMMENTED** [review]
just a few minor formatting suggestions

**Review by kayx23 — APPROVED** [review]

File: docs/en/latest/tutorials/how-to-avoid-race-condition-when-scaling-down.md [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+metadata:
+  name: example-pod
+spec:
+  containers:
+  - name: web-server
+    image: web-server:latest
+    lifecycle:
+      preStop:
+        exec:
+          command: ["/bin/sh", "-c", "sleep 5 && kill -SIGTERM $(cat /var/run/webserver.pid)"]
```
Comment (Sn0rt) [inline docs/en/latest/tutorials/how-to-avoid-race-condition-when-scaling-down.md:61]: This user case is not concise enough and misleading. I think just writing sleep 5 is enough. It means that sleep 5 gives the apisix ingress controller enough time to process the control plane changes.

**Comment by Revolyssup** [comment]
@Sn0rt I applied the suggestion. PTAL

**Review by Sn0rt — APPROVED** [review]

**Review by AlinsRan — APPROVED** [review]

**Pull request merged**